### ISSUE TITLE
[Feature] Run Ginkgo tests with Ginkgo CLI

### ISSUE THREAD

**Issue by MortalHappiness** [ray-project/kuberay#2658]
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Currently, Ginkgo tests are run with `go test`, and we cannot run a single test, which is annoying. We should use the Ginkgo CLI to run Ginkgo tests.

### Implementation Details

- Please search for `Describe(` using your IDE to identify which files contain Ginkgo tests.  
- Try running the tests locally using the Ginkgo CLI and fix any errors that occur.  
- Attempt to run a single test with the Ginkgo CLI; it should pass. Update the test to intentionally make it fail, then run it again with the Ginkgo CLI to ensure it fails as expected.  
- Update the configurations in the `.github/` and `.buildkite/` folders to use the Ginkgo CLI instead of `go test` for running Ginkgo tests.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by nadongjun** [comment]
@MortalHappiness Hi! I’d like to take this on—please assign it to me if possible. Thanks!

**Comment by dentiny** [comment]
Does `go test -run` work here?
Eg: https://stackoverflow.com/questions/26092155/just-run-single-test-instead-of-the-whole-suite

**Comment by MortalHappiness** [comment]
> Does `go test -run` work here? Eg: https://stackoverflow.com/questions/26092155/just-run-single-test-instead-of-the-whole-suite

No, it only works for regular Golang tests, not Ginkgo tests.

**Comment by dentiny** [comment]
> No, it only works for regular Golang tests, not Ginkgo tests.

Hmm would this work?
https://stackoverflow.com/questions/68008485/how-to-run-a-single-spec-using-ginkgo-test-framework

**Comment by nadongjun** [comment]
I've tested running specific Ginkgo tests using `go test -v -ginkgo.focus "Describe"`, and it works as expected. 

While `go test -ginkgo.focus` also supports running individual tests, transitioning to the Ginkgo CLI offers additional advantages, such as output formatting, support for parallel execution, custom reporters, and better debugging information.

Given these benefits, do you think switching from go test to the Ginkgo CLI would be a good move?

**Comment by MortalHappiness** [comment]
> do you think switching from go test to the Ginkgo CLI would be a good move?

Yes. I still think it's better to switch to use Ginkgo CLI, because it has more features.

**Comment by nadongjun** [comment]
Upon reviewing the issue, it was confirmed that a conflict occurs when Ginkgo tests are mixed with standard Go tests. (When running Ginkgo --focus, all tests defined in the standard go test are also executed.) This happens because, to use Ginkgo with go test, all test code must be written in the Ginkgo format. The flags -ginkgo.focus and -ginkgo.skip only work with tests written in the Ginkgo framework.

Conflicts can arise when Ginkgo tests are mixed with standard Go tests, and to resolve this, all tests need to be standardized to the Ginkgo format. Converting all tests to the Ginkgo format will require considerable time and effort, making it a long-term task. For further reference on similar issues, you can check [Ginkgo Issue #358](https://github.com/onsi/ginkgo/issues/358).

**Comment by MortalHappiness** [comment]
Could we use `ginkgo convert` as mentioned in that issue?

**Comment by nadongjun** [comment]
@MortalHappiness 
Regarding the use of ginkgo convert, it has been deprecated starting from Ginkgo v2. This means it is no longer recommended for use in newer versions of Ginkgo. If you are working with Ginkgo v2 or later, it would be best to avoid ginkgo convert and explore alternative approaches.

@andrewsykim 
Sure, I've added a new make target that runs the Ginkgo tests with the option to pass in a focus string. This is intended for local development, so there's no need to run it through CI. The target requires the TEST_NAME and the path to the test folder to be specified. If go test is available, it runs the test with the provided TEST_NAME. If not, it falls back to running the Ginkgo tests with the focus string. Here's how it works:

### Script:
```sh
test-target: WHAT ?= ./controllers/ray/
test-target: TEST_NAME ?= ""
test-target: ENVTEST_K8S_VERSION ?= 1.24.2
test-target: manifests fmt vet envtest
	@go test -run $(TEST_NAME) $(WHAT) -count=1 | tee >(if grep -q "no tests to run"; then \
		echo "No matching Go tests found. Running Ginkgo tests…"; \
		KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) —bin-dir $(LOCALBIN) -p path)" ginkgo —focus "$(TEST_NAME)" $(WHAT); \
	fi) 
```

### For ginkgo test:
```sh
make test-target TEST_NAME=RayClusterwithinvalidNumOfHosts WHAT=./controllers/ray/
```

### For go test:
```sh
make test-target TEST_NAME=TestCreateRayJobSubmitterIfNeed
```

If this approach looks reasonable, I will proceed to create a PR.

**Comment by MortalHappiness** [comment]
Hi @nadongjun. Are you still working on this issue?

**Comment by nadongjun** [comment]
Hi @MortalHappiness,

I've tried working on this, but I haven't been able to make much progress, and it's currently stuck. I think it would be better to unassign myself so someone else can take it over.
